FUSETOOLS-2267 - Karaf 4.x server adapter

lhein opened the pull request:
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

lhein commented:
This is a draft PR. It is still not fully working. Server start and stop is working fine but you are not able to connect to JMX. Deployment is fully untested for now.

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/plugin.xml:
```diff
@@ -346,32 +414,62 @@
        			<facet id="jst.web"  version="2.2,2.3,2.4,2.5,3.0"/>
        			<facet id="jst.java" version="1.3,1.4,5.0,6.0,1.7,1.8"/>
        			<facet id="jst.utility" version="1.0"/>
-       			<facet id="jst.camel" version="2.15"/>
```
don't we have a bug here in current version of the product?

lhein replied:
yes, pretty much.

apupier replied:
What do you think to report it and create a PR separately?
As we don't know exactly when this PR will be merged (as amount of work to do on it is not well defined as it is "test and discover"), it will make this PR simpler and ensure that the fix is available and tested sooner.

lhein replied:
addressed in #1089

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/plugin.xml:
```diff
@@ -346,32 +414,62 @@
+	</adapter>
+	<adapter>
+      <runtime-component id="org.fusesource.ide.karaf.runtime.component" version="4.1"/>
+      <factory class="org.fusesource.ide.server.karaf.core.runtime.classpath.KarafClasspathProviderAdapterFactory"/>
```
as it is the same factory fo r4.0 and 4.1, shouldn't it be simple to have a single one for 4+ version?

lhein replied:
Yeah, good point. We should merge them into a single adapter 4+ for now.

lhein replied:
when looking at other server adapter plugins in JBT they also separate the versions. I'd like to keep it separated then.

apupier replied:
ok

apupier commented:
Can you put the JIRA number on each commits please? Otherwise after the merge, we will no more know to which Jira issue it corresponds

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/server/subsystems/Karaf4xStartupLaunchConfigurator.java:
```diff
@@ -0,0 +1,49 @@
+	@Override
+	protected void findJars(IPath path, List<Object> cp) {
+		File[] libs = path.toFile().listFiles( (File pathname) -> pathname.isDirectory() || (pathname.isFile() && pathname.getName().toLowerCase().endsWith(".jar")));
+		for (File lib : libs) {
```
I think libs can be null

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/server/subsystems/Karaf4xStartupLaunchConfigurator.java:
```diff
@@ -0,0 +1,49 @@
+		for (File lib : libs) {
+			IPath p = path.append(lib.getName());
+			if (lib.isFile()) {
+				if (lib.getName().toLowerCase().indexOf("karaf")!=-1 || lib.getPath().toLowerCase().indexOf("boot")!=-1) {
```
lib.getName().toLowerCase(). can be extracted

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/util/KarafUtils.java:
```diff
@@ -513,6 +513,7 @@ public static String getJMXConnectionURL(IServer server) {
 				url = url.substring(pos + placeHolder.length() + 3);
 			}
 		}
+		if (retVal.length()<1) retVal.append(url);
```
I would prefer to have it on 2 different lines

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.ui/plugin.xml:
```diff
@@ -44,6 +54,16 @@
             typeIds="org.fusesource.ide.karaf.server.30">
       </fragment>
+      <fragment
+            class="org.fusesource.ide.server.karaf.ui.runtime.v2x.KarafServerWizardFragment2x"
```
strange to have the 2x in the name although it is for 3 and 4 version too

lhein replied:
I agree. Maybe we should consider a refactoring in future. (out of scope of this PR)

apupier replied:
have you created a JIRA task for it?

lhein replied:
https://issues.jboss.org/browse/FUSETOOLS-2528

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/server/Karaf4xServerDelegate.java:
```diff
@@ -0,0 +1,48 @@
+ * @author lheinema
+ *
+ */
+public class Karaf4xServerDelegate extends Karaf3xServerDelegate {
```
wouldn't it be better to extract the common part of those classes in an abstract one and have specific bit of code for 3x and 4x in separated children class?

lhein replied:
I agree. Maybe we should consider a refactoring in future. (out of scope of this PR)

apupier replied:
I don't get why it is outside of the scope of this PR? the Karaf4xServerDelegate is introduced in this PR and the current structure is introducing technical debt, why not doing it right the first time?

lhein replied:
because it requires refactoring of the Karaf3xServerDelegate too. Lets create a refactor jira and tackle it separately.

lhein replied:
https://issues.jboss.org/browse/FUSETOOLS-2528

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/runtime/KarafRuntimeLocator.java:
```diff
@@ -144,7 +144,8 @@ public IRuntimeWorkingCopy getRuntimeFromDir(File dir, IProgressMonitor monitor)
 				if( type.equals(KarafBeanProvider.KARAF_2x) || 
-					type.equals(KarafBeanProvider.KARAF_3x) ) {
+					type.equals(KarafBeanProvider.KARAF_3x) ||
+					type.equals(KarafBeanProvider.KARAF_4x)) {
```
better to invert equals and it might avoid the need to check for nullity of type just before

lhein commented:
@apupier this PR is solely for FUSETOOLS-2267 and I mentioned that number in the initial commit message. It will be suashmerged anyways so why wasting time on labeling each single commit in this PR.

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/runtime/KarafRuntimeDelegate.java:
```diff
@@ -89,6 +89,10 @@ public IStatus validate() {
         } else if (version != null && version.trim().startsWith("3.0")) {
             if (!id.toLowerCase().endsWith("karaf.runtime.30")) return new Status(Status.ERROR, Activator.PLUGIN_ID, "Runtime type not compatible with found version...");
+        } else if (version != null && version.trim().startsWith("4.0")) {
+            if (!id.toLowerCase().endsWith("karaf.runtime.40")) return new Status(Status.ERROR, Activator.PLUGIN_ID, "Runtime type not compatible with found version...");
```
might worth to provide the found version and the runtime type in the message logged

apupier replied:
might worth extract this duplicated String "Runtime type not compatible with found version..."

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/publish/jmx/KarafBundleMBeanPublishBehaviour.java:
```diff
@@ -159,4 +160,14 @@ public boolean canHandle(MBeanServerConnection mbsc) {
 	}
+	
+	protected String getBundleSymbolicName(CompositeData cd) {
+		String bsn = null;
```
I think the null assignation is useless

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/publish/jmx/KarafBundleMBeanPublishBehaviour.java:
```diff
@@ -159,4 +160,14 @@ public boolean canHandle(MBeanServerConnection mbsc) {
+	protected String getBundleSymbolicName(CompositeData cd) {
+		String bsn = null;
+		try {
+			bsn = cd.get("Symbolic Name").toString();
```
isn't the differecne between Karaf 2-3 and 4? if yes, what do you thinkto have a specific class for each case? or at least to extract methods mentioning that one is for 4- and the other for 4+?

lhein replied:
the point is I don't know if thats true. Could be also that latest 2.4 or 3 versions changed that too.

lhein replied:
And its just this one logic change so far. If we get more then it makes probably sense.

apupier commented on jmx/plugins/org.fusesource.ide.jmx.karaf/src/org/fusesource/ide/jmx/karaf/navigator/osgi/CompositeDataPropertySource.java:
```diff
@@ -82,6 +82,14 @@ public CompositeDataPropertySource(CompositeData cd) {
 	}
+	
+	private boolean isNumberProperty(String className) {
+		return 	className.equalsIgnoreCase("java.lang.byte") ||
```
better to invert equal

lhein commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/runtime/KarafRuntimeLocator.java:
```diff
@@ -92,12 +85,7 @@ public void search(File folder, IRuntimeSearchListener listener, int recursionLe
-					return pathname.isDirectory();
-				}
-			});
+			files = folder.listFiles( (File pathname) -> pathname.isDirectory() );
```
I am not really seeing any advantage in following that rule in this specific case.

apupier replied:
folder.listFiles( (File pathname) -> pathname.isDirectory() );

vs

folder.listFiles(File::isDirectory);

the advantage is that it is shorter

lhein replied:
I find that less readable but thats probably personal taste.

apupier commented:
> @apupier this PR is solely for FUSETOOLS-2267 and I mentioned that number in the initial commit message. It will be suashmerged anyways so why wasting time on labeling each single commit in this PR.

so you mean that you want to use the "squash and merge". The downside of it is that we have a single commit at end, it might be interesting to have some history for instance for different goal of commit, here it would be interesting to have in different commits:
- Karaf 4.x server adapter initial version
- switch to ssh shell utf8 encoding for 4.x
- allow startup of snapshot versions of the runtime for Karaf 4.x
- Support undeploy for Karaf 4.x
- repaired bundles property page to show number values like the identifier correctly

apupier commented:
not sure that it is only for Karaf 4.x but worth checking:
there is no icon for Karaf Launcher so there is a blank space when launched in debug mode:
![image](https://user-images.githubusercontent.com/1105127/30173486-1e3cda00-93f8-11e7-94f5-8a180788511e.png)

lhein commented:
> @apupier this PR is solely for FUSETOOLS-2267 and I mentioned that number in the initial commit message. It will be suashmerged anyways so why wasting time on labeling each single commit in this PR.
>
>so you mean that you want to use the "squash and merge". The downside of it is that we have a single commit at end, it might be interesting to have some history for instance for different goal of commit, here it would be interesting to have in different commits:
>- Karaf 4.x server adapter initial version
>- switch to ssh shell utf8 encoding for 4.x
>- allow startup of snapshot versions of the runtime for Karaf 4.x
>- Support undeploy for Karaf 4.x
>- repaired bundles property page to show number values like the identifier correctly

We at one point agreed to keep the commit history clean and less noisy and I am pretty sure we at that point agreed on using the Squash & Merge functionality of GitHub. You and @bfitzpat might correct me if I remember that wrong.

apupier commented:
> We at one point agreed to keep the commit history clean and less noisy and I am pretty sure we at that point agreed on using the Squash & Merge functionality of GitHub. You and @bfitzpat might correct me if I remember that wrong.

I think we agreed on that before "rebase and merge" was available and when it appeared decided to move to "rebase and merge"

lhein commented:
> - I cannot use the 4.2-SNAPSHOT version because it is marked as not compatible

That version is not valid. I don't see any karaf download having that version. It is always X.Y.Z at least.

apupier commented:
> I cannot use the 4.2-SNAPSHOT version because it is marked as not compatible
> That version is not valid. I don't see any karaf download having that version. It is always X.Y.Z at least.

my bad i wasn't enough precise, the version is 4.2.0-SNAPSHOT as you can see in the screenshot, the 4.2-SNAPSHOT is only the name that I gave

lhein commented:
> EDIT: this already an issue for other versions: https://issues.jboss.org/browse/FUSETOOLS-2527 but it might require to add an entry for Karaf 4.x too in this PR

Nope...we reuse the 2xLaunchType.

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/bean/ServerBeanTypeKaraf4x.java:
```diff
@@ -0,0 +1,106 @@
+		protected static boolean isIntegratedKaraf(File location) {
+			File libFolder = new File(location + File.separator + "lib");
+			File[] files = libFolder.listFiles( (File dir, String name) -> name.toLowerCase().endsWith("-version.jar") );
+			return files.length > 0;
```
files can be null

apupier commented on servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/server/Karaf4xServerDelegate.java:
```diff
@@ -0,0 +1,48 @@
+	@Override
+	public IStatus validate() {
+		// check if the folder exists and the karaf.jar is in place
+		IPath rtLoc = getServer().getRuntime().getLocation();
```
NPE in case the runtime doesn't exists anymore on filesystem:
<details>
!ENTRY org.eclipse.jface 4 2 2017-09-11 14:22:13.743
!MESSAGE Problems occurred when invoking code from plug-in: "org.eclipse.jface".
!STACK 0
java.lang.NullPointerException
	at org.fusesource.ide.server.karaf.core.server.Karaf4xServerDelegate.validate(Karaf4xServerDelegate.java:32)
	at org.fusesource.ide.server.karaf.core.server.subsystems.Karaf2xLaunchController.canStart(Karaf2xLaunchController.java:105)
	at org.jboss.ide.eclipse.as.wtp.core.server.behavior.ControllableServerBehavior.canStart(ControllableServerBehavior.java:269)
	at org.eclipse.wst.server.core.internal.Server.canStart(Server.java:1655)
	at org.eclipse.wst.server.ui.internal.view.servers.StartAction.accept(StartAction.java:138)
	at org.eclipse.wst.server.ui.internal.view.servers.AbstractServerAction.selectionChanged(AbstractServerAction.java:85)
	at org.eclipse.wst.server.ui.internal.view.servers.StartAction.selectionChanged(StartAction.java:186)
	at org.eclipse.ui.actions.SelectionProviderAction.selectionChanged(SelectionProviderAction.java:144)
	at org.eclipse.jface.viewers.Viewer$1.run(Viewer.java:155)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:44)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:173)
	at org.eclipse.jface.viewers.Viewer.fireSelectionChanged(Viewer.java:152)
	at org.eclipse.jface.viewers.StructuredViewer.updateSelection(StructuredViewer.java:2180)
	at org.eclipse.jface.viewers.StructuredViewer.handleSelect(StructuredViewer.java:1223)
	at org.eclipse.ui.navigator.CommonViewer.handleSelect(CommonViewer.java:463)
	at org.eclipse.jface.viewers.StructuredViewer$4.widgetSelected(StructuredViewer.java:1252)
	at org.eclipse.jface.util.OpenStrategy.fireSelectionEvent(OpenStrategy.java:240)
	at org.eclipse.jface.util.OpenStrategy.access$4(OpenStrategy.java:235)
	at org.eclipse.jface.util.OpenStrategy$1.handleEvent(OpenStrategy.java:400)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:86)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4428)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1079)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4238)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3817)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1044)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:680)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:594)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:151)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1499)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1472)
</details>

lhein commented:
> you mentioned that it was written "Apache Tomcat" in the list and adjusted to "Apache Karaf" but I have only "Tomcat" so might be best to keep Karaf only under Apache category:

Weird...my fault. Will correct it.

lhein commented:
testPR

lhein commented:
I rebased with master and fixed the issue with the Identifier log message. However there are still messages left regarding "headers" column which is currently not handled in a good way but that is the state since the very beginning of the karaf server adapter. We should track that in a separate issue.

apupier commented on core/plugins/org.fusesource.ide.foundation.ui/src/org/fusesource/ide/foundation/ui/properties/NumberPropertyDescriptor.java:
```diff
@@ -0,0 +1,39 @@
+public class NumberPropertyDescriptor extends PropertyDescriptor implements ReturnType {
+	
+	/**
+	 * creates a property descriptor for boolean properties
```
NumberpropertyDescriptor for a Boolean properties? seems something is not correct somewhere ^^

apupier commented on core/plugins/org.fusesource.ide.foundation.ui/src/org/fusesource/ide/foundation/ui/properties/NumberPropertyDescriptor.java:
```diff
@@ -0,0 +1,39 @@
+ * @author lheinema
+ *
+ */
+public class NumberPropertyDescriptor extends PropertyDescriptor implements ReturnType {
```
ReturnType is a parameterized type, we can write something like:

public class NumberPropertyDescriptor extends PropertyDescriptor implements ReturnType<String> {
	
	/**
	 * creates a property descriptor for number properties
	 * 
	 * @param id	the id
	 * @param displayName	the display name
	 */
	public NumberPropertyDescriptor(Object id, String displayName) {
		super(id, displayName);
	}

	@Override
	public Class<String> getReturnType() {
		return String.class;
	}

}
